pkgtools/pkg_alternatives/files/pkg_alternatives.8
.\" $NetBSD: pkg_alternatives.8,v 1.7 2020/01/10 12:20:50 leot Exp $
.\"
.\" pkg_alternatives - Generic wrappers for programs with similar interfaces
.\" Copyright (c) 2005 Julio M. Merino Vidal <jmmv@NetBSD.org>
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Neither the name of The NetBSD Foundation nor the names of its
.\"    contributors may be used to endorse or promote products derived
.\"    from this software without specific prior written permission.
.\" 3. Neither the name of author nor the names of its contributors may
.\"    be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
.\" ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
.\" TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
.\" BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd January 10, 2020
.Dt PKG_ALTERNATIVES 8
.Os
.Sh NAME
.Nm pkg_alternatives
.Nd generic wrappers for programs with similar interfaces
.Sh SYNOPSIS
.Nm
.Op Fl gsw
.Op Fl p Ar prefix
.Cm action
.Ar arg1 ... argN
.Sh DESCRIPTION
.Nm
is a tool to manage the
.Em alternatives system
provided by The
.Nx
Packages Collection, also known as pkgsrc.
It creates, configures, and destroys generic wrappers used to run
programs with similar interfaces.
.Pp
Consider, for example, the
.Xr vi 1
editor.
Both
.Xr nvi 1
and
.Xr vim 1
provide very similar functionality, although neither of them is named
.Pa vi .
In this situation, it may be useful to have a
.Em wrapper
in place of the generic name which points to one of the multiple
.Em alternatives
available.
.Pp
The alternatives are classified according to the package they belong to,
forming
.Em groups
of alternatives.
In other words, a concrete package contains a list of all the wrappers it
affects by providing alternatives to them.
This makes things easier for the end user and/or the administrator who has
to deal with them, as they can manually select a group (package) rather than a
bunch of wrappers.
.Pp
Wrappers are identified by their filename relative to the installation
prefix, i.e.,
.Pa @PREFIX@
if the
.Fl p
option is not used.
In the example above, the wrapper could be identified by the
.Pa bin/vi
string.
The identifier is the string used as the
.Dq wrapper
argument in all the actions that require it.
.Pp
The generic wrapper scans a list of available alternatives and tries to
execute them in order; the first one that succeeds is the one used for
that run.
This list of alternatives is read from multiple configuration files.
Each one contains a list of possible alternatives, one per line, with
optional arguments to them.
Lines starting with
.Sq #
are considered comments and are ignored.
.Pp
The following configuration files are read, in strict order, for each
wrapper (replace the
.Dq wrapper
word with the wrapper's absolute file name):
.Bl -tag -width XXXX
.It Pa ~/.pkg_alternatives/wrapper
This is called the
.Em user configuration file
and is only read when not running as
.Dq a privileged user .
Otherwise, tools such as
.Xr sudo 8
could be used to execute any program in the system.
This file can be freely edited by the user, either by hand or by using
the
.Cm manual
action described below.
.It Pa @CONFDIR@/wrapper
This is called the
.Em system configuration file
and is read if found.
This file can be freely edited by the administrator, either by hand or
by using the
.Cm manual
action described below.
.It Pa @DBDIR@/wrapper
This is called the
.Em database configuration file
and is always read.
This file must not be edited by hand; packages providing alternatives
will take care to (un)register themselves from them when (de)installed.
.El
.Ss Options
The following options are available:
.Bl -tag -width XpXprefixX
.It Fl g
Operate on groups of wrappers rather than individual files.
This is the opposite of
.Fl w
and is currently the default behavior.
Affects the behavior of all actions.
.It Fl p Ar prefix
Set installation prefix.
This affects where wrappers and their manual pages are looked for.
The prefix defaults to
.Pa @PREFIX@
if this flag is not given.
.It Fl s
Run in silent mode: no output except for errors.
.It Fl w
Operate on individual wrappers rather than on groups.
This is the opposite of
.Fl g .
Affects the behavior of all actions.
.El
.Ss Actions in group mode
The following table describes each available action and its behavior when
working in group mode.
All these actions are at a higher level than the same actions in wrapper
mode.
They always end up using the later at some point, so you should also
read the next section to be aware of the exact effect of each command.
.Bl -tag -width XXXX
.It Cm auto Ar package
Removes any manual configuration from each wrapper associated with the given
package.
This means that all affected wrappers are then free to choose whichever
alternative they prefer.
.It Cm destroy
Removes the alternative database, found in
.Pa @DBDIR@ ,
and all its associated wrappers and manual pages.
This action is intended to be used by the
.Nm
package.
.It Cm list
Lists which of the installed packages provide alternatives.
Any of the packages shown by this command can then be fed back to the
.Cm auto
and
.Cm manual
actions.
.It Cm manual Ar package
Manually selects all the alternatives that belong to
.Ar package
to be the default for their respective wrappers.
.It Cm rebuild
Rebuilds the alternatives database, found in
.Pa @DBDIR@ ,
based on the contents of the package database, usually available in
.Pa @PKG_DBDIR@ .
Basically, it scans the later looking for packages with an
.Pa +ALTERNATIVES
file in them, and, for each of those, the
.Cm register
action is called with the appropriate file name.
This action is intended to be used by the
.Nm
package or in case of database corruption.
.It Cm register Ar package wrapper alternative arguments
Registers a new
.Ar alternative
for the given
.Ar wrapper
in the specified
.Ar package .
If the package does not exist in the database, it is created.
.Pp
This action must not be used directly; packages providing alternatives
will take care to execute it at installation time.
.It Cm status Ar package
For each wrapper that belongs to
.Ar package ,
shows which alternative will be used by it in the next run.
It also displays all available candidates for each of them.
.It Cm unregister Ar package
Removes the
.Ar package
from the database.
All alternatives associated to it are also removed.
.Pp
This action must not be used directly; packages providing alternatives
will take care to execute it at deinstallation time.
.El
.Ss Actions in wrapper mode
The following table describes each available action and its behavior when
working in wrapper mode.
Note that these actions work at a very low level as they are used to manage
wrappers and alternative commands directly.
In most situations, you will want to use these actions in group mode.
.Bl -tag -width XXXX
.It Cm auto Ar wrapper
Removes any manual configuration created for the given
.Ar wrapper .
That is, if running as
.Dq a privileged user ,
the system configuration file is deleted; otherwise, the user configuration
file is removed.
The effect of this action is that the wrapper is then free to choose any
alternative it wants.
.It Cm manual Ar wrapper alternative arguments
Manually selects the
.Ar alternative
for the given
.Ar wrapper .
If running as
.Dq a privileged user ,
the system configuration file is modified; otherwise, the user configuration
file is changed.
The effect of this action is that the wrapper will try to use your preferred
alternatives, regardless of what is installed on the system.
.It Cm register Ar wrapper alternative arguments
Registers a new
.Ar alternative
for the given
.Ar wrapper .
If the wrapper did not exist before, it is created.
You may optionally pass several
.Ar arguments
to the
.Ar alternative
program.
.Pp
This action should not be used directly; packages providing alternatives
will take care to execute it at installation time.
.It Cm status Ar wrapper
Shows which alternative will be used by the
.Ar wrapper
in the next run.
It also displays all available candidates for it.
.It Cm unregister Ar wrapper alternative
Removes the
.Ar alternative
from the given
.Ar wrapper .
If there are no more alternatives available, the wrapper is removed.
.Pp
This action should not be used directly; packages providing alternatives
will take care to execute it at deinstallation time.
.El
.Ss Filtering wrappers
.Nm
lets you choose which wrappers you want on your system and which ones should
simply be ignored.
This is accomplished by a filter matched against every wrapper, which is
defined in the
.Pa @CONFDIR@/filter.conf
file.
.Pp
A filter is composed of multiple entries.
Each entry contains an action and a regular expression, separated by a
.Em single space .
The action can be either
.Dq accept
or
.Dq ignore .
The former specifies that, if the regular expression is matched against a
wrapper name, processing should stop and the wrapper should be created.
The later is exactly the opposite: if the name matches the expression, the
wrapper is ignored and processing stops.
.Pp
Please note that, after modifying the filter configuration file, the
wrappers database
.Em must be rebuilt
using the
.Dq rebuild
action.
Otherwise your changes will take no effect.
.Sh ENVIRONMENT
.Bl -tag -width PKG_DBDIR
.It Ev PKG_DBDIR
Location of the package database directory.
Defaults to
.Pa @PKG_DBDIR@ .
.El
.Sh FILES
.Bl -tag -width XXXX
.It Pa ~/.pkg_alternatives/
User-specific configuration directory.
.It Pa @CONFDIR@/
System-wide configuration directory.
.It Pa @CONFDIR@/filter.conf
Wrapper filter.
.It Pa @DBDIR@/
System-wide configuration database.
.El
.Sh DIAGNOSTICS
.Nm
exists 0 on success and 1 if an error occurred.
.Sh EXAMPLES
.Ss Managing wrapper groups
The following command tells all Vim related wrappers (which include
.Xr ex 1 ,
.Xr vi 1
and
.Xr view 1 )
to always prefer Vim in favour of any other program:
.Bd -literal -offset indent
# pkg_alternatives manual vim
.Ed
.Pp
And the following command reverts the previous change, configuring the
affected wrappers to use whichever alternative is available:
.Bd -literal -offset indent
# pkg_alternatives auto vim
.Ed
.Ss Managing individual wrappers
Suppose that you want to use Sun's Java 1.5 by default for all Java-related
wrappers, except for
.Xr appletviewer 1 ,
because you want to use Kaffe in that case.
The following commands do this, by first selecting Sun's Java 1.5 and later
overriding the exact wrapper to refer to Kaffe:
.Bd -literal -offset indent
# pkg_alternatives manual sun-jre15
# pkg_alternatives manual sun-jdk15
# pkg_alternatives -w manual bin/appletviewer \\
  /usr/pkg/bin/kaffe-appletviewer
.Ed
.Ss Applying filters
Let us consider a very typical situation: you have just installed a Python
interpreter and you want the
.Pa bin/python
wrapper to be created, but you do not want to pollute your system with any
other wrapper (such as those coming from Vim).
You can achieve this by using a filter that first accepts the Python wrapper
and then ignores everything else.
The following lines could be added to
.Pa @CONFDIR@/filter.conf
to achieve this:
.Bd -literal -offset indent
accept ^bin/python$
ignore .*
.Ed
.Pp
Don't forget to run the following command after doing the above changes:
.Bd -literal -offset indent
# pkg_alternatives rebuild
.Ed
.Sh SEE ALSO
.Xr pkg_add 1 ,
.Xr pkg_delete 1
.Sh HISTORY
The
.Nm
utility first appeared in pkgsrc-2005Q1.
.Pp
This utility was inspired by the alternatives system found in the Debian
operating system.
.Sh AUTHORS
.An Julio M. Merino Vidal Aq Mt jmmv@NetBSD.org


pkgtools/pkg_alternatives/files/pkg_alternatives.sh
#!@SH@
#
# $NetBSD: pkg_alternatives.sh,v 1.7 2012/06/13 15:35:32 jperkin Exp $
#
# pkg_alternatives - Generic wrappers for programs with similar interfaces
# Copyright (c) 2005 Julio M. Merino Vidal <jmmv@NetBSD.org>
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in
#    the documentation and/or other materials provided with the
#    distribution.
# 3. Neither the name of author nor the names of its contributors may
#    be used to endorse or promote products derived from this software
#    without alternative prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
# ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
# TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
# BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
#

Conf_Dir=@CONFDIR@@PREFIX@
Data_Dir=@DATADIR@
Db_Dir=@DBDIR@@PREFIX@
Filter=
Filter_Read=no
Prefix=@PREFIX@
Prog_Name=${0##*/}
Verbose=yes

: ${PKG_DBDIR:=@PKG_DBDIR@}
: ${PKGMANDIR:=@PKGMANDIR@}

# -------------------------------------------------------------------------

# action_auto_package package
#
# Configures the given package to be in automatic mode.  I.e., it removes
# local customizations for all related wrappers, so that only the files in
# the database are taken into account.
#
action_auto_package() {
    validate_args auto ${#} -eq 1
    validate_package ${1}
    pkg=${PKG_DBDIR}/${1}*/+ALTERNATIVES

    set -- $(cat ${pkg} | sed -e 's# #__dE/lImIt/Er__#g')
    while [ ${#} -gt 0 ]; do
        action_auto_wrapper ${1%%__dE/lImIt/Er__*}
        shift
    done
}

# -------------------------------------------------------------------------

# action_auto_wrapper wrapper
#
# Configures the given wrapper to be in automatic mode.  I.e., it removes
# local customizations so that only the files in the database are taken
# into account.
#
# wrapper is a path relative to prefix.
#
action_auto_wrapper() {
    validate_args auto ${#} -eq 1
    validate_wrapper ${1} yes

    conf=$(get_my_config)/${1}
    if [ -f ${conf} ]; then
        info "removing configuration from \`${conf}'"
        rm -f ${conf}
        rmdir_p ${conf%/*}
    else
        info "no existing configuration for \`${1}'; nothing to be done"
    fi
}

# -------------------------------------------------------------------------

# action_destroy_package
#
# Destroys the alternatives database and removes all installed wrappers.
#
action_destroy_package() {
    validate_args list ${#} -eq 0

    [ -d ${Db_Dir} ] || return 0

    wrappers=$(cd ${Db_Dir} ; find . -type f)

    for w in ${wrappers}; do
        wrapper=$(echo ${w} | sed -e 's|^./||')
        manpage=$(get_manpage ${wrapper})
        info "removing wrapper \`${wrapper}'"
        rm -f ${Prefix}/${wrapper}
        info "removing manual page \`${manpage}'"
        rm -f ${manpage}
    done

    info "removing contents of \`@DBDIR@'"
    rm -rf @DBDIR@/* 2>/dev/null
    rmdir_p @DBDIR@
}

# -------------------------------------------------------------------------

# action_destroy_wrapper
#
# Unavailable action.
#
action_destroy_wrapper() {
    err "the \`destroy' action cannot be used in wrapper mode"
}

# -------------------------------------------------------------------------

# action_list_package
#
# Lists all available packages that provide alternatives.
#
action_list_package() {
    validate_args list ${#} -eq 0

    info "looking for alternatives in \`${PKG_DBDIR}'"
    for d in ${PKG_DBDIR}/*; do
        [ -f ${d}/+ALTERNATIVES ] && echo ${d##${PKG_DBDIR}/}
    done
}

# -------------------------------------------------------------------------

# action_list_wrapper
#
# Unavailable action.
#
action_list_wrapper() {
    err "the \`list' action cannot be used in wrapper mode"
}

# -------------------------------------------------------------------------

# action_manual_package package
#
# Configures the given package to be the preferred alternative for all the
# wrappers associated to it.
#
action_manual_package() {
    validate_args manual ${#} -eq 1
    validate_package ${1}
    pkg=${PKG_DBDIR}/${1}*/+ALTERNATIVES

    set -- $(cat ${pkg} | sed -e 's# #__dE/lImIt/Er__#g')
    while [ ${#} -gt 0 ]; do
        action_manual_wrapper $(echo ${1} | sed -e 's#__dE/lImIt/Er__# #g')
        shift
    done
}

# -------------------------------------------------------------------------

# action_manual_wrapper wrapper alternative arguments
#
# Configures the given wrapper to use the specified alternative.
# If running as the administrator, the system-wide settings are changed;
# otherwise, the user-specific settings are affected.
#
# wrapper is a path relative to prefix.
# alternative is an absolute path.
#
action_manual_wrapper() {
    validate_args manual ${#} -ge 2
    validate_wrapper ${1} yes
    validate_alternative ${2}

    conf=$(get_my_config)/${1}
    info "modifying configuration from \`${conf}'"
    mkdir_p ${conf%/*}
    shift
    echo "${*}" >${conf}
}

# -------------------------------------------------------------------------

# action_rebuild_package
#
# Rebuilds the alternatives database from the package database and also
# reinstalls the wrappers (so that filter.conf takes effect).
#
action_rebuild_package() {
    validate_args list ${#} -eq 0

    action_destroy_package
    info "looking for alternatives in \`${PKG_DBDIR}'"
    for d in ${PKG_DBDIR}/*; do
        if [ -f ${d}/+ALTERNATIVES ]; then
            action_register_package ${d}/+ALTERNATIVES
        fi
    done
}

# -------------------------------------------------------------------------

# action_rebuild_wrapper
#
# Unavailable action.
#
action_rebuild_wrapper() {
    err "the \`rebuild' action cannot be used in wrapper mode"
}

# -------------------------------------------------------------------------

# action_register_package file
#
# Registers all wrapper/alternative pairs listed in the file.  This is
# supposed to be an +ALTERNATIVES file found inside a package in PKG_DBDIR.
# Each line should follow the semantics expected by action_register_wrapper.
#
action_register_package() {
    validate_args register ${#} -eq 1

    set -- $(cat ${1} | sed -e 's# #__dE/lImIt/Er__#g')
    while [ ${#} -gt 0 ]; do
        action_register_wrapper $(echo ${1} | sed -e 's#__dE/lImIt/Er__# #g')
        shift
    done
}

# -------------------------------------------------------------------------

# action_register_wrapper wrapper alternative arguments
#
# Registers a new alternative for the given wrapper in the database.
# The wrapper and its manual page are created if they do not exist.
#
action_register_wrapper() {
    validate_args register ${#} -ge 2
    validate_wrapper ${1} no
    validate_alternative ${2}

    wbase=${1}; shift
    alt=${1}; shift
    args=${*}

    filter ${wbase} || return

    wabs=${Prefix}/${wbase}
    manpage=$(get_manpage ${wbase})
    dbconf=${Db_Dir}/${wbase}
    sysconf=${Conf_Dir}/${wbase}

    if [ ! -f ${dbconf} ]; then
        info "initializing database entry for \`${wbase}'"
        mkdir_p ${dbconf%/*}
        touch ${dbconf%/*}
    fi
    info "registering alternative \`${alt}' for wrapper \`${wbase}'"
    if [ -n "${args}" ]; then
        echo "${alt} ${args}" >>${dbconf}
    else
        echo "${alt}" >>${dbconf}
    fi

    if [ ! -f ${wabs} ]; then
        info "creating wrapper \`${wbase}'"
        mkdir_p ${wabs%/*}
        sed \
            -e "s|__SH__|@SH@|g" \
            -e "s|__CONF_FILE__|${sysconf}|g" \
            -e "s|__CREATOR__|${Prog_Name}|g" \
            -e "s|__DB_FILE__|${dbconf}|g" \
            -e "s|__IS_ROOT_CMD__|@IS_ROOT_CMD@|g" \
            -e "s|__WRAPPER__|${wabs}|g" \
            <${Data_Dir}/wrapper.sh >${wabs}
        chmod +x ${wabs}
        mkdir_p ${manpage%/*}
        sed -e "s|__CONF_FILE__|${sysconf}|g" \
            -e "s|__CREATOR__|${Prog_Name}|g" \
            -e "s|__DB_FILE__|${dbconf}|g" \
            -e "s|__PREFIX__|${Prefix}|g" \
            -e "s|__SECTION__|$(get_manpage_sect ${wbase})|g" \
            -e "s|__TITLE__|$(echo ${wbase##*/} | tr a-z A-Z)|g" \
            -e "s|__WRAPPER__|${wbase##*/}|g" \
            -e "s|__WRAPPERBASE__|${wbase}|g" \
            <${Data_Dir}/wrapper.man >${manpage}
    fi
}

# -------------------------------------------------------------------------

# action_status_package package
#
# Shows the status of each wrapper associated to the given package.
#
action_status_package() {
    validate_args status ${#} -eq 1
    validate_package ${1}
    pkg=${PKG_DBDIR}/${1}*/+ALTERNATIVES

    set -- $(cat ${pkg} | sed -e 's# #__dE/lImIt/Er__#g')
    while [ ${#} -gt 0 ]; do
        action_status_wrapper ${1%%__dE/lImIt/Er__*}
        shift
    done
}

# -------------------------------------------------------------------------

# action_status_wrapper wrapper
#
# Shows the current status for the given wrapper.
#
action_status_wrapper() {
    validate_args status ${#} -eq 1
    validate_wrapper ${1} yes

    wbase=${1}
    dbconf=${Db_Dir}/${wbase}
    sysconf=${Conf_Dir}/${wbase}
    userconf=~/.pkg_alternatives${Prefix}/${wbase}

    @IS_ROOT_CMD@ && userconf=
    alts=$(cat ${userconf} ${sysconf} ${dbconf} 2>/dev/null | grep -v '^#' | \
           sed -e 's# #__dE/lImIt/Er__#g')

    found=
    for a in ${alts}; do
        prog=${a%%__dE/lImIt/Er__*}
        if [ -x ${prog} ]; then
            found=$(echo ${a} | sed -e 's#__dE/lImIt/Er__# #g')
            break
        fi
    done

    [ -n "${found}" ] ||
        err "the wrapper \`${wbase}' exists but has no valid alternatives"

    echo "\`${wbase}' points to \`${found}'"
    for a in $(echo ${alts} | tr ' ' '\n' | sort | uniq); do
        echo "    candidate: $(echo ${a} | sed -e 's#__dE/lImIt/Er__# #g')"
    done
}

# -------------------------------------------------------------------------

# action_unregister_package file
#
# Unregisters the given package (given by its +ALTERNATIVES file) and
# removes all associated alternatives from their respective wrappers.
#
action_unregister_package() {
    validate_args unregister ${#} -eq 1

    set -- $(cat ${1} | sed -e 's# #__dE/lImIt/Er__#g')
    while [ ${#} -gt 0 ]; do
        action_unregister_wrapper $(echo ${1} | sed -e 's#__dE/lImIt/Er__# #g')
        shift
    done
}

# -------------------------------------------------------------------------

# action_unregister_wrapper wrapper alternative arguments
#
# Unregisters an alternative for the given wrapper.  The lookup is done
# based on the alternative name and its arguments.  I.e., there must be
# an exact match.  If no alternatives remain after the removal, the wrapper
# and its manpage are removed.
#
action_unregister_wrapper() {
    validate_args unregister ${#} -ge 2
    validate_wrapper ${1} yes

    wbase=${1}; shift
    alt=${1}; shift
    args=${*}

    manpage=$(get_manpage ${wbase})
    dbconf=${Db_Dir}/${wbase}

    if [ -n "${args}" ]; then
        grep -v "^${alt} ${args}$" <${dbconf} >${dbconf}.new
    else
        grep -v "^${alt}$" <${dbconf} >${dbconf}.new
    fi
    if cmp -s ${dbconf} ${dbconf}.new; then
        rm -f ${dbconf}.new
        err "unknown alternative \`${alt}' for wrapper \`${wbase}'"
    fi
    mv ${dbconf}.new ${dbconf}

    if ! grep '^/' ${dbconf} >/dev/null; then
        info "no more alternatives for \`${wbase}'; removing"
        rm -f ${dbconf} ${Prefix}/${wbase} ${manpage}
        rmdir_p ${dbconf%/*}
        rmdir_p ${Prefix}/${wbase%/*}
        rmdir_p ${manpage%/*}
    fi
}

# -------------------------------------------------------------------------

# err message
#
# Prints the given error message and exist with an error code.
#
err() {
    echo "${Prog_Name}: ${*}" 1>&2
    exit 1
}

# -------------------------------------------------------------------------

# filter wrapper
#
# Reads the configuration filter from CONFDIR/filter.conf and, for the
# given wrapper, returns whether it is accepted or ignored.
#
filter() {
    [ ! -f @CONFDIR@/filter.conf ] && return 0

    if [ ${Filter_Read} = no ]; then
        Filter=$(cat @CONFDIR@/filter.conf | grep -v '^#' | sed -e 's# #__dE/lImIt/Er__#g')
        Filter_Read=yes
    fi

    [ -z "${Filter}" ] && return 0

    for f in ${Filter}; do
        what=${f%%__dE/lImIt/Er__*}
        case ${what} in
            accept)
                name=$(echo ${f#*__dE/lImIt/Er__} | sed -e 's#__dE/lImIt/Er__# #g')
                if echo ${1} | grep "${name}" >/dev/null; then
                    info "filter accepts \`${1}'"
                    return 0
                fi
                ;;
            ignore)
                name=$(echo ${f#*__dE/lImIt/Er__} | sed -e 's#__dE/lImIt/Er__# #g')
                if echo ${1} | grep "${name}" >/dev/null; then
                    info "filter ignores \`${1}'"
                    return 1
                fi
                ;;
            *)
                warn "unknown filter type \`${what}'; ignoring"
                ;;
        esac
    done

    true
}

# -------------------------------------------------------------------------

# get_my_config
#
# Prints the configuration directory for the current user.  If root, this
# is the system-wide configuration directory.  Otherwise, it is the user's
# personal directory.
#
get_my_config() {
    if @IS_ROOT_CMD@; then
        echo ${Conf_Dir}
    else
        echo ~/.pkg_alternatives${Prefix}
    fi
}

# -------------------------------------------------------------------------

# get_manpage wrapper
#
# Prints the absolute path to the manual page associated to wrapper.
# Does no error checking; validate_wrapper has to be called before this
# function to ensure wrapper is a valid name.
#
get_manpage() {
    sect=$(get_manpage_sect ${1})
    echo ${Prefix}/${PKGMANDIR}/man${sect}/${1##*/}.${sect}
}

# -------------------------------------------------------------------------

# get_manpage_sect wrapper
#
# Prints the section number that will be used by the wrapper's manpage.
# Does no error checking; validate_wrapper has to be called before this
# function to ensure wrapper is a valid name.
#
get_manpage_sect() {
    case ${1%/*} in
        bin) echo 1 ;;
        libexec|sbin) echo 8 ;;
    esac
}

# -------------------------------------------------------------------------

# info message
#
# Prints the given informative message if running in verbose mode.
#
info() {
    [ ${Verbose} = yes ] && echo "${Prog_Name}: ${*}"
}

# -------------------------------------------------------------------------

# mkdir_p directory
#
mkdir_p() {
    if [ ! -d ${1} ]; then
        cnt=2
        max=$(($(echo ${1} | tr '/' ' ' | wc -w | awk '{ print $1; }') + 1))
        while [ ${cnt} -le ${max} ]; do
            mkdir $(echo ${1} | cut -d '/' -f -${cnt}) 2>/dev/null
            cnt=$((${cnt} + 1))
        done
        [ -d ${1} ] || err "cannot create directory \`${1}'"
    fi
}

# -------------------------------------------------------------------------

# usage
#
# Shows an usage message and exits the program with an error condition.
#
usage() {
    cat 1>&2 <<EOF
Usage: ${Prog_Name} [options] action [arg1 ... argN]

Available options:
    -g        Select group mode (default).
    -p dir    Set installation prefix.
    -s        Run in silent mode.
    -w        Select wrapper mode.

Available actions in group mode:
    auto destroy list manual rebuild register status unregister

Available actions in wrapper mode:
    auto manual register status unregister

See pkg_alternatives(8) for more information.
EOF
    exit 1
}

# -------------------------------------------------------------------------

# rmdir_p directory
#
rmdir_p() {
    if [ -d ${1} ]; then
        cnt=$(($(echo ${1} | tr '/' ' ' | wc -w | awk '{ print $1; }') + 1))
        while [ ${cnt} -gt 1 ]; do
            rmdir $(echo ${1} | cut -d '/' -f -${cnt}) 2>/dev/null
            cnt=$((${cnt} - 1))
        done
    fi
}

# -------------------------------------------------------------------------

# validate_alternative name
#
# Verifies that the given program name is valid to be used as an
# alternative.
#
validate_alternative() {
    echo ${1} | grep '^/' >/dev/null || \
        err "the alternative \`${1}' is not an absolute file name"
    [ -x ${1} ] || \
        err "the alternative \`${1}' is not an executable"
}

# -------------------------------------------------------------------------

# validate_args action argcount operator expected
#
# Verifies that the number of arguments passed to the specified action
# are correct according to its semantics.  Prints an error message if
# incorrect.
#
validate_args() {
    [ ${2} ${3} ${4} ] || \
        err "incorrect number of arguments for the \`${1}' action"
}

# -------------------------------------------------------------------------

# validate_package name
#
# Verifies that the specified package is valid.
#
validate_package() {
    if [ ! -f ${PKG_DBDIR}/${1}/+ALTERNATIVES ]; then
        cnt=$(cd ${PKG_DBDIR} && ls -d ${1}-[0-9]* 2>/dev/null | wc -l | \
              awk '{ print $1; }')
        if [ "${cnt}" -eq 0 ]; then
            err "the package \`${1}' is not known"
        elif [ "${cnt}" -gt 1 ]; then
            err "multiple matches found for \`${1}'"
        fi

        # Reached this point, there is a single match for the package,
        # but it may still be wrong (i.e., no alternatives for it).
        [ ! -f ${PKG_DBDIR}/${1}*/+ALTERNATIVES ] && \
            err "no alternatives defined for the \`${1}' package"
    fi
}

# -------------------------------------------------------------------------

# validate_wrapper name exists
#
# Verifies that the specified wrapper is valid, i.e., if it belongs to a
# known directory.  Also, if the exists parameter is yes, the function
# ensures that the wrapper already exists (as well as its configuration
# file in the database).
#
validate_wrapper() {
    case ${1%/*} in
        bin|libexec|sbin)
            if [ ${2} = yes ]; then
                [ -x ${Prefix}/${1} ] || \
                    err "the wrapper \`${1}' does not exist or is not" \
                        "executable"
                [ -f ${Db_Dir}/${1} ] || \
                    err "the database configuration \`${Db_Dir}/${1}'" \
                        "does not exist"
            fi
            ;;
        *)
            err "wrapper name \`${1}' is invalid"
            ;;
    esac
}

# -------------------------------------------------------------------------

# warn message
#
# Shows a warning message.
#
warn() {
    echo "${Prog_Name}: ${*}" 1>&2
}

# -------------------------------------------------------------------------

# main
#
# Main program code.  Does argument parsing and executes the required
# action.
#
main() {
    args=$(getopt gp:sw ${*})
    [ ${?} -eq 0 ] || usage
    set -- ${args}
    what=package
    while [ ${#} -gt 0 ]; do
        case ${1} in
            -g)
                what=package
                ;;
            -p)
                Prefix=$2; shift
                Conf_Dir=@CONFDIR@${Prefix}
                Db_Dir=@DBDIR@${Prefix}
                ;;
            -s)
                Verbose=no
                ;;
            -w)
                what=wrapper
                ;;
            --)
                shift; break
                ;;
        esac
        shift
    done

    if [ ${#} -eq 0 ]; then
        usage
    fi

    action=${1}; shift
    case ${action} in
        auto|destroy|list|manual|rebuild|register|status|unregister)
            action_${action}_${what} "${@}"
            ;;
        *)
            err "unknown action \`${action}'"
            ;;
    esac

    return 0
}

main "${@}"

# vim: syntax=sh:expandtab:shiftwidth=4:softtabstop=4
